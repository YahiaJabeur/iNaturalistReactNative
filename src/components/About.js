// @flow

import Button from "components/SharedComponents/Buttons/Button";
import { View } from "components/styledComponents";
import { t } from "i18next";
import type { Node } from "react";
import React from "react";
import { Alert } from "react-native";
import { getBuildNumber, getSystemName, getVersion } from "react-native-device-info";
import Mailer from "react-native-mail";

import { logFilePath } from "../../react-native-logs.config";
import PlaceholderText from "./PlaceholderText";
import ViewWithFooter from "./SharedComponents/ViewWithFooter";

const AboutScreen = ( ): Node => {
  const appVersion = getVersion( );
  const buildVersion = getBuildNumber( );
  const device = getSystemName( );

  const emailParams = {
    subject: `iNat RN ${device} Logs (version ${appVersion} - ${buildVersion})`,
    recipients: ["help+mobile@inaturalist.org"]
  };

  const setAttachment = ( ) => ( {
    path: logFilePath,
    mimeType: "txt"
  } );

  const openEmailWithLogsAttached = ( ) => {
    Mailer.mail( {
      ...emailParams,
      isHTML: true,
      attachments: [setAttachment( )]
    }, ( error, event ) => {
      Alert.alert(
        error,
        event
      );
    } );
  };

  return (
    <ViewWithFooter>
      <PlaceholderText text={`app version: ${appVersion} (${buildVersion})`} />
      <View className="mt-16 mx-5">
        <Button
          level="primary"
          onPress={openEmailWithLogsAttached}
          text={t( "EMAIL-DEBUG-LOGS" )}
        />
      </View>
    </ViewWithFooter>
  );
};

export default AboutScreen;
