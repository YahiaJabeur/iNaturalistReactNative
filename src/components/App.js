// @flow

import AsyncStorage from "@react-native-async-storage/async-storage";
import { fetchUserMe } from "api/users";
import { getUserId, signOut } from "components/LoginSignUp/AuthenticationService";
import i18next from "i18next";
import RootDrawerNavigator from "navigation/rootDrawerNavigation";
import { RealmContext } from "providers/contexts";
import type { Node } from "react";
import React, { useEffect } from "react";
import useAuthenticatedQuery from "sharedHooks/useAuthenticatedQuery";
import useCurrentUser from "sharedHooks/useCurrentUser";

const { useRealm } = RealmContext;

const App = ( ): Node => {
  const realm = useRealm( );
  const currentUser = useCurrentUser( );

  // fetch current user from server and save to realm in useEffect
  // this is used for changing locale
  const {
    data: user
  } = useAuthenticatedQuery(
    ["fetchUserMe"],
    optsWithAuth => fetchUserMe( { }, optsWithAuth )
  );

  useEffect( ( ) => {
    const checkForSignedInUser = async ( ) => {
      const userId = await getUserId( );
      // check to see if this is a fresh install of the app
      // if it is, delete realm file when we sign the user out of the app
      // this handles the case where a user deletes the app, then reinstalls
      // and expects to be signed out with no previously saved data
      const alreadyLaunched = await AsyncStorage.getItem( "alreadyLaunched" );
      let deleteRealm = false;
      if ( !alreadyLaunched ) {
        deleteRealm = true;
        await AsyncStorage.setItem( "alreadyLaunched", "true" );
      }
      if ( !userId ) {
        await signOut( { deleteRealm } );
      }
    };

    checkForSignedInUser( );
  }, [] );

  useEffect( ( ) => {
    if ( user ) {
      realm?.write( ( ) => {
        realm?.create( "User", user, "modified" );
      } );
      if ( !user.locale ) { return; }
      i18next.changeLanguage( user.locale, err => {
        if ( err ) { throw new Error( err ); }
      } );
    }
  }, [user, realm, currentUser] );

  return (
    <RootDrawerNavigator />
  );
};

export default App;
