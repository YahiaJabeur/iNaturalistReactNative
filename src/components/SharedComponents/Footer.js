// @flow

import * as React from "react";
import { Pressable, Text, View } from "react-native";
import { useNavigation } from "@react-navigation/native";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

import { viewStyles } from "../../styles/sharedComponents/footer";
// import CameraOptionsModal from "../Camera/CameraOptionsModal";
// import Modal from "./Modal";
import CameraOptionsButton from "./Buttons/CameraOptionsButton";

const Footer = ( ): React.Node => {
  // const [showModal, setModal] = React.useState( false );

  // const openModal = React.useCallback( ( ) => setModal( true ), [] );
  // const closeModal = React.useCallback( ( ) => setModal( false ), [] );

  const navigation = useNavigation( );
  const toggleSideMenu = ( ) => navigation.openDrawer( );
  const navToObsList = ( ) => navigation.navigate( "my observations" );
  // const navToCameraOptions = ( ) => openModal( );
  const navToExplore = ( ) => navigation.navigate( "explore stack" );
  const navToNotifications = ( ) => navigation.navigate( "notifications" );

  return (
    <>
      {/* <Modal
        showModal={showModal}
        closeModal={closeModal}
        modal={<CameraOptionsModal closeModal={closeModal} />}
      /> */}
      <View style={[viewStyles.row, viewStyles.shadow]}>
        <Pressable onPress={toggleSideMenu} accessibilityRole="link">
          <Icon name="menu" size={30} />
        </Pressable>
        <Pressable  onPress={navToExplore} accessibilityRole="link">
          <Icon name="web" size={30} />
        </Pressable>
        <CameraOptionsButton buttonType="footer" />
        {/* <Pressable onPress={navToCameraOptions} accessibilityRole="link">
          <Text>camera</Text>
        </Pressable> */}
        <Pressable onPress={navToObsList} accessibilityRole="link">
          <Icon name="account" size={30} />
        </Pressable>
        <Pressable onPress={navToNotifications} accessibilityRole="link">
          <Icon name="bell" size={30} />
        </Pressable>
      </View>
    </>
  );
};

export default Footer;
