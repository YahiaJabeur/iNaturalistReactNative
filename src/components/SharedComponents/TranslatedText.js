// @flow

import type { Node } from "react";
import React from "react";
import { useTranslation } from "react-i18next";
import { Text } from "react-native";

type Props = {
  text: string,
  textStyle?: string,
  count?: number
}

const TranslatedText = ( { text, textStyle, count }: Props ): Node => {
  const { t } = useTranslation( );

  return <Text className={textStyle}>{t( text, { count } )}</Text>;
};

export default TranslatedText;
