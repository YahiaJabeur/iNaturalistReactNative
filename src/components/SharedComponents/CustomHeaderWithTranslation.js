// @flow

import { Text } from "components/styledComponents";
import type { Node } from "react";
import React from "react";
import { useTranslation } from "react-i18next";

type Props = {
  headerText: string
}

const CustomHeaderWithTranslation = ( { headerText }: Props ): Node => {
  const { t } = useTranslation( );

  return (
    <Text className="text-2xl">{t( headerText )}</Text>
  );
};

export default CustomHeaderWithTranslation;
