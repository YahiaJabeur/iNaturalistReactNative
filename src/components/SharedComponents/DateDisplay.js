// @flow

import { Body4 } from "components/SharedComponents";
import { View } from "components/styledComponents";
import * as React from "react";
import { useTranslation } from "react-i18next";
import IconMaterial from "react-native-vector-icons/MaterialIcons";
import { formatObsListTime } from "sharedHelpers/dateAndTime";

type Props = {
  dateTime: string | typeof undefined
};

const DateDisplay = ( {
  dateTime
}: Props ): React.Node => {
  const { t } = useTranslation( );
  const displayTime = ( ) => {
    if ( dateTime ) {
      return formatObsListTime( dateTime );
    }
    return t( "no time given" );
  };

  return (
    <View className="flex flex-row items-center">
      <IconMaterial name="watch-later" size={15} />
      <Body4 className="text-darkGray ml-[5px]">
        {displayTime( )}
      </Body4>
    </View>
  );
};

export default DateDisplay;
