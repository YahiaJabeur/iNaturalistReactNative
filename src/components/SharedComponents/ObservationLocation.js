// @flow

import { Body4 } from "components/SharedComponents";
import { View } from "components/styledComponents";
import * as React from "react";
import { useTranslation } from "react-i18next";
import IconMaterial from "react-native-vector-icons/MaterialIcons";

type Props = {
  observation: Object
};

const ObservationLocation = ( { observation }: Props ): React.Node => {
  const { t } = useTranslation();

  let locationName = observation?.place_guess;

  if (
    !locationName
    // Check for undefined or null, Not 0
    && observation?.latitude != null
    && observation?.longitude != null
  ) {
    locationName = `${observation.latitude}, ${observation.longitude}`;
  } else if ( !locationName ) {
    locationName = t( "Missing Location" );
  }

  return (
    <View className="flex flex-row items-center">
      <IconMaterial name="location-pin" size={15} />
      <Body4 className="text-darkGray ml-[5px]">
        {locationName || t( "Missing-Location" )}
      </Body4>
    </View>
  );
};

export default ObservationLocation;
