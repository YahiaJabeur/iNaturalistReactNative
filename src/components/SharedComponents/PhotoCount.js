// @flow

import { Body3 } from "components/SharedComponents";
import { View } from "components/styledComponents";
import * as React from "react";
import Svg, { Path } from "react-native-svg";
import { viewStyles } from "styles/sharedComponents/photoCount";
import colors from "styles/tailwindColors";

type Props = {
  count: number,
  size?: number,
  shadow?: boolean,
};

const PhotoCount = ( { count, size, shadow }: Props ): React.Node => {
  if ( count === 0 ) {
    return null;
  }

  let photoCount = count;
  if ( count > 99 ) {
    photoCount = 99;
  }

  return (
    <View
      style={[{ height: size, width: size }, shadow && viewStyles.shadow]}
      testID="photo-count"
    >
      <Svg
        height={size}
        width={size}
        viewBox="0 0 24 24"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <Path
          fillRule="nonzero"
          clipRule="evenodd"
          d="M4 5.818a4 4 0 00-4 4V20a4 4 0 004 4h10.182a4 4 0 004-4V9.818a4 4 0 00-4-4z"
          fill={colors.white}
        />
        <Path
          // eslint-disable-next-line
          d="M15.364 3.636h-9.53A4 4 0 019.818 0H20a4 4 0 014 4v10.182a4 4 0 01-3.636 3.984v-9.53a5 5 0 00-5-5z"
          fill={colors.white}
          clipRule="evenodd"
          fillRule="nonzero"
        />
        <Body3>
          {photoCount}
        </Body3>
      </Svg>
    </View>
  );
};

export default PhotoCount;

PhotoCount.defaultProps = {
  size: 24,
  shadow: false
};
