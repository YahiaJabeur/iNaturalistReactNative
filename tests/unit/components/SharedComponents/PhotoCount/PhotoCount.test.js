import { render, screen } from "@testing-library/react-native";
import { PhotoCount } from "components/SharedComponents";
import React from "react";

describe( "PhotosCount", () => {
  it( "should render correctly", () => {
    render( <PhotoCount count={10} /> );

    const photoCount = screen.queryByTestId( "photo-count" );

    expect( photoCount ).toBeTruthy();
  } );

  it( "should not render when photo count equal to 0", () => {
    render( <PhotoCount count={0} /> );

    const photoCount = screen.queryByTestId( "photo-count" );

    expect( photoCount ).toBeNull();
  } );

  it( "should show the correct value", () => {
    render( <PhotoCount count={14} /> );

    expect( screen ).toMatchSnapshot();
  } );

  it( "should show photo count value equal to 99", () => {
    render( <PhotoCount count={200} /> );

    expect( screen ).toMatchSnapshot();
  } );
} );
