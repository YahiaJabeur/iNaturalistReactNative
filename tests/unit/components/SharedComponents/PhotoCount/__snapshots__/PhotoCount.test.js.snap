// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosCount should show photo count value equal to 99 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "height": 24,
          "width": 24,
        },
        false,
      ],
    ]
  }
  testID="photo-count"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={24}
    bbWidth={24}
    fill="none"
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        Array [
          "fill",
        ]
      }
    >
      <RNSVGPath
        clipRule={0}
        d="M4 5.818a4 4 0 00-4 4V20a4 4 0 004 4h10.182a4 4 0 004-4V9.818a4 4 0 00-4-4z"
        fill={4294967295}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
      <RNSVGPath
        clipRule={0}
        d="M15.364 3.636h-9.53A4 4 0 019.818 0H20a4 4 0 014 4v10.182a4 4 0 01-3.636 3.984v-9.53a5 5 0 00-5-5z"
        fill={4294967295}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4282729797}
        font={
          Object {
            "fontSize": "10",
            "textAnchor": "middle",
          }
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            "38%",
          ]
        }
        y={
          Array [
            "77%",
          ]
        }
      >
        <RNSVGTSpan
          content="99"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`PhotosCount should show the correct value 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "height": 24,
          "width": 24,
        },
        false,
      ],
    ]
  }
  testID="photo-count"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={24}
    bbWidth={24}
    fill="none"
    focusable={false}
    height={24}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 24,
          "width": 24,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      fill={null}
      propList={
        Array [
          "fill",
        ]
      }
    >
      <RNSVGPath
        clipRule={0}
        d="M4 5.818a4 4 0 00-4 4V20a4 4 0 004 4h10.182a4 4 0 004-4V9.818a4 4 0 00-4-4z"
        fill={4294967295}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
      <RNSVGPath
        clipRule={0}
        d="M15.364 3.636h-9.53A4 4 0 019.818 0H20a4 4 0 014 4v10.182a4 4 0 01-3.636 3.984v-9.53a5 5 0 00-5-5z"
        fill={4294967295}
        fillRule={1}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
      <RNSVGText
        content={null}
        dx={Array []}
        dy={Array []}
        fill={4282729797}
        font={
          Object {
            "fontSize": "10",
            "textAnchor": "middle",
          }
        }
        propList={
          Array [
            "fill",
          ]
        }
        rotate={Array []}
        x={
          Array [
            "38%",
          ]
        }
        y={
          Array [
            "77%",
          ]
        }
      >
        <RNSVGTSpan
          content="14"
          dx={Array []}
          dy={Array []}
          font={Object {}}
          rotate={Array []}
          x={Array []}
          y={Array []}
        />
      </RNSVGText>
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
